File: 3630.partition-array-for-maximum-xor-and-and.go
#
# @lc app=leetcode id=3630 lang=golang
#
# [3630] Partition Array for Maximum XOR and AND
#

# @lc code=start
type State struct {
	xa uint32
	xc uint32
	ab uint32
}

func generate(half []int) []State {
	res := make([]State, 0, 1 << 16)
	var dfs func(int, uint32, uint32, uint32)
	dfs = func(i int, x1 uint32, x2 uint32, andb uint32) {
		if i == len(half) {
			res = append(res, State{x1, x2, andb})
			return
		}
		nu := uint32(half[i])
		dfs(i+1, x1^nu, x2, andb)
		dfs(i+1, x1, x2^nu, andb)
		dfs(i+1, x1, x2, andb & nu)
	}
	dfs(0, 0, 0, ^uint32(0))
	return res
}

func maximizeXorAndXor(nums []int) int64 {
	n := len(nums)
	mid := n / 2
	lefts := generate(nums[:mid])
	rights := generate(nums[mid:])
	maxV := int64(0)
	ones := ^uint32(0)
	for _, l := range lefts {
		for _, r := range rights {
			abT := l.ab & r.ab
			if abT == ones {
				abT = 0
			}
			xaT := l.xa ^ r.xa
			xcT := l.xc ^ r.xc
			v := int64(xaT) + int64(xcT) + int64(abT)
			if v > maxV {
				maxV = v
			}
		}
	}
	return maxV
}
# @lc code=end

File: 3630.partition-array-for-maximum-xor-and-and.java
#
# @lc app=leetcode id=3630 lang=java
#
# [3630] Partition Array for Maximum XOR and AND
#

# @lc code=start
class Solution {
  public long maximizeXorAndXor(int[] nums) {
    int n = nums.length;
    int mid = (n + 1) / 2;
    int lenL = mid;
    int lenR = n - mid;
    int numL = 1;
    for (int i = 0; i < lenL; i++) numL *= 3;
    int numR = 1;
    for (int i = 0; i < lenR; i++) numR *= 3;
    int[] xorAL = new int[numL];
    int[] xorCL = new int[numL];
    int[] andBL = new int[numL];
    boolean[] hasBL = new boolean[numL];
    for (int state = 0; state < numL; state++) {
      int temp = state;
      int xa = 0, xc = 0, ab = ~0;
      boolean hb = false;
      for (int k = 0; k < lenL; k++) {
        int ass = temp % 3;
        temp /= 3;
        int val = nums[k];
        if (ass == 0) {
          xa ^= val;
        } else if (ass == 2) {
          xc ^= val;
        } else {
          ab &= val;
          hb = true;
        }
      }
      xorAL[state] = xa;
      xorCL[state] = xc;
      andBL[state] = ab;
      hasBL[state] = hb;
    }
    int[] xorAR = new int[numR];
    int[] xorCR = new int[numR];
    int[] andBR = new int[numR];
    boolean[] hasBR = new boolean[numR];
    for (int state = 0; state < numR; state++) {
      int temp = state;
      int xa = 0, xc = 0, ab = ~0;
      boolean hb = false;
      for (int k = 0; k < lenR; k++) {
        int ass = temp % 3;
        temp /= 3;
        int val = nums[mid + k];
        if (ass == 0) {
          xa ^= val;
        } else if (ass == 2) {
          xc ^= val;
        } else {
          ab &= val;
          hb = true;
        }
      }
      xorAR[state] = xa;
      xorCR[state] = xc;
      andBR[state] = ab;
      hasBR[state] = hb;
    }
    long ans = 0;
    for (int i = 0; i < numL; i++) {
      int xal = xorAL[i];
      int xcl = xorCL[i];
      int abl = andBL[i];
      boolean hbl = hasBL[i];
      for (int j = 0; j < numR; j++) {
        int xar = xorAR[j];
        int xcr = xorCR[j];
        int abr = andBR[j];
        boolean hbr = hasBR[j];
        long abv;
        if (hbl && hbr) {
          abv = (long) abl & abr;
        } else if (hbl) {
          abv = abl;
        } else if (hbr) {
          abv = abr;
        } else {
          abv = 0;
        }
        long xat = (long) (xal ^ xar);
        long xct = (long) (xcl ^ xcr);
        long val = xat + xct + abv;
        if (val > ans) {
          ans = val;
        }
      }
    }
    return ans;
  }
}
# @lc code=end